### ISSUE TITLE
Support seek on the files opened from read-only zipfiles

### ISSUE THREAD

**Issue by rkhwaja** [PyFilesystem/pyfilesystem2#98]

**Comment by willmcgugan** [comment]
I haven't confirmed this, but I suspect the [file objects](https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile.open) returned from `zipfile` just aren't seekable.

It's unfortunate but the only workaround would be to copy the file to a temp file first, which would incur a performance penalty.

Unless I can come up with a way of implementing seek that doesn't impact anything else, I may have to close this as 'wontfix'. Open to suggestions for a way to make that work...

**Comment by althonos** [comment]
Some seek operations can be emulated for no cost. I'll explore this issue a
bit and come back with my findings.

**Comment by althonos** [comment]
This is even better than what I expected : after studying the source of [`zipfile.py`](https://github.com/python/cpython/blob/master/Lib/zipfile.py), this is what I understood:

when opening a file in a zipfile, it will create a `_SharedFile` object that locks on reading, and does not support seeking, but maps the virtual position in the file to the actual position in the Zip file stream. So we could bypass the `_SharedFile` and implement locking ourselves in a custom `ZipExtFile`, which would be similar to what I did when I implemented `ISOFS`. That way, we could have real seeking as long as the underlying Zip file stream supports streaming.
